src/coreclr/vm/peimagelayout.cpp

```diff
@@ -758,6 +681,457 @@ FlatImageLayout::FlatImageLayout(PEImage* pOwner)
+    HANDLE                 Process,
+    PVOID                  BaseAddress,
+    ULONG64                Offset,
+    SIZE_T                 ViewSize,
+    ULONG                  AllocationType,
+    ULONG                  PageProtection,
+    MEM_EXTENDED_PARAMETER* ExtendedParameters,
+    ULONG                  ParameterCount);
+
+MapViewOfFile3Fn pMapViewOfFile3 = NULL;
+
+static bool HavePlaceholderAPI()
+{
+    const MapViewOfFile3Fn INVALID_ADDRESS_CENTINEL = (MapViewOfFile3Fn)1;
+    if (pMapViewOfFile3 == INVALID_ADDRESS_CENTINEL)
+    {
+        return false;
+    }
+
+    if (pMapViewOfFile3 == NULL)
+    {
+        HMODULE hm = LoadLibraryW(_T("kernelbase.dll"));
+        if (hm != NULL)
+        {
+            pVirtualAlloc2 = (VirtualAlloc2Fn)GetProcAddress(hm, "VirtualAlloc2");
+            pMapViewOfFile3 = (MapViewOfFile3Fn)GetProcAddress(hm, "MapViewOfFile3");
+
+            FreeLibrary(hm);
+        }
+
+        if (pMapViewOfFile3 == NULL || pVirtualAlloc2 == NULL)
+        {
+            pMapViewOfFile3 = INVALID_ADDRESS_CENTINEL;
+            return false;
+        }
+    }
+
+    return true;
+}
+
+static PVOID AllocPlaceholder(PVOID BaseAddress, SIZE_T Size)
+{
+    return pVirtualAlloc2(
+        ::GetCurrentProcess(),
+        BaseAddress,
+        Size,
+        MEM_RESERVE | MEM_RESERVE_PLACEHOLDER,
+        PAGE_NOACCESS,
+        NULL,
+        0);
+}
+
+static PVOID MapIntoPlaceholder(
+    HANDLE                 FileMapping,
+    ULONG64                FromOffset,
+    PVOID                  ToAddress,
+    SIZE_T                 ViewSize,
+    ULONG                  PageProtection
+)
+{
+    return pMapViewOfFile3(FileMapping, ::GetCurrentProcess(), ToAddress, FromOffset, ViewSize, MEM_REPLACE_PLACEHOLDER, PageProtection, NULL, 0);
+}
+
+static PVOID CommitIntoPlaceholder(
+    PVOID                  ToAddress,
+    SIZE_T                 Size,
+    ULONG                  PageProtection
+)
+{
+    return pVirtualAlloc2(::GetCurrentProcess(), ToAddress, Size, MEM_COMMIT | MEM_RESERVE | MEM_REPLACE_PLACEHOLDER, PageProtection, NULL, 0);
+}
+
+static PVOID SplitPlaceholder(
+    PVOID& placeholderStart,
+    PVOID   placeholderEnd,
+    SIZE_T  size
+)
+{
+    _ASSERTE((char*)placeholderStart + size <= placeholderEnd);
+
+    if ((char*)placeholderStart + size < placeholderEnd)
+    {
+        if (!VirtualFree(placeholderStart, size, MEM_RELEASE | MEM_PRESERVE_PLACEHOLDER))
+        {
+            return NULL;
+        }
+    }
+
+    PVOID result = placeholderStart;
+    placeholderStart = (char*)placeholderStart + size;
+
+    return result;
+}
+
+static SIZE_T OffsetWithinPage(SIZE_T addr)
+{
+    return addr & (GetOsPageSize() - 1);
+}
+
+static SIZE_T RoundToPage(SIZE_T size, SIZE_T offset)
+{
+    size_t result = size + OffsetWithinPage(offset);
+    _ASSERTE(result >= size);
+    return ROUND_UP_TO_PAGE(result);
+}
+
+void* FlatImageLayout::LoadImageByMappingParts(SIZE_T* m_imageParts) const
```

Comment:
Would there be a way to factor this lengthy method into a few smaller ones (e.g. by defining a separate method to deal with the mapping of a single section) and ideally get rid of the numerous <code>goto</code> statements? I don't recall when was the last time I saw so many in a single CoreCLR native function :-).

Reply:
This started as a Windows version of PAL `MAPMapPEFile` method. It losely follows the same logic - go through header/sections and map them and bail via `goto` if something unexpected found in the file. There are many things that may go wrong at different stages. Avoiding goto here leads to duplication of the "failed" sequence.

Reply:
The style is common. It is just typically done via macros like `IF_FAIL_GO`, so noone sees the actual `goto`.

I'll see what can be done here, but it could the case where `goto FAILED;` is better than alternatives.